fog/view/server/src/db_fetcher.rs

```diff
@@ -268,60 +268,34 @@ impl<DB: RecoveryDb + Clone + Send + Sync + 'static> DbFetcherThread<DB> {
         let mut has_more_work = false;
 
         // See whats the next block number we need to load for each invocation we are aware of.
-        let (ingestable_ranges, missing_block_ranges) = {
-            let shared_state = self.shared_state();
-            (
-                shared_state.ingestable_ranges.clone(),
-                shared_state.missing_block_ranges.clone(),
-            )
-        };
+        let ingress_keys = self.shared_state().ingress_keys.clone();
 
-        let next_block_index_per_invocation_id = self.block_tracker.next_blocks(&ingestable_ranges);
+        let next_block_index_per_ingress_key = self.block_tracker.next_blocks(&ingress_keys);
 
         log::trace!(
             self.logger,
             "load_block_data next_blocks: {:?}",
-            next_block_index_per_invocation_id
+            next_block_index_per_ingress_key
         );
 
-        for (ingest_invocation_id, block_index) in next_block_index_per_invocation_id.into_iter() {
-            // Do not attempt to fetch blocks explicitly marked as missing.
-            if BlockTracker::is_missing_block(&missing_block_ranges, block_index) {
-                self.block_tracker
-                    .block_processed(ingest_invocation_id, block_index);
-
-                // This ensures we do not have holes in the blocks processed by the enclave thread.
-                self.shared_state().fetched_records.push(FetchedRecords {
-                    ingest_invocation_id,
-                    block_index,
-                    records: vec![],
-                });
-
-                continue;
-            }
-
+        for (ingress_key, block_index) in next_block_index_per_ingress_key.into_iter() {
             // Attempt to load data for the next block.
             let get_tx_outs_by_block_result = {
                 let _metrics_timer = counters::GET_TX_OUTS_BY_BLOCK_TIME.start_timer();
                 self.db
-                    .get_tx_outs_by_block(&ingest_invocation_id, block_index)
+                    .get_tx_outs_by_block_and_key(ingress_key, block_index)
             };
 
             match get_tx_outs_by_block_result {
-                Ok(tx_outs) => {
-                    let num_tx_outs = tx_outs.len();
-
-                    // NOTE: This makes a very nuanced and important assumption, which is that
-                    // ingest always produces data ETxOutRecords for each block it has processed,
-                    // EVEN if it actually found NO matches.
-                    // Based on that assumption, tx_outs will be empty only when ingest has not yet
-                    // ingested the block (and wrote the results into the database).
-                    if !tx_outs.is_empty() {
+                Ok(maybe_tx_outs) => {
```

Comment:
I think this will be more nicely written as 3 clauses in the match block:
```
Ok(Some(tx_outs)) => ...
Ok(None) => ...
Err(err) => ...
```

Reply:
oh yeah, that is better